File: src/client/java/net/greenjab/fixedminecraft/FixedMinecraftClient.java
package net.greenjab.fixedminecraft;

import net.fabricmc.api.ClientModInitializer;
import net.fabricmc.fabric.api.client.rendering.v1.hud.HudElementRegistry;
import net.fabricmc.fabric.api.resource.ResourceManagerHelper;
import net.fabricmc.fabric.api.resource.ResourcePackActivationType;
import net.fabricmc.loader.api.FabricLoader;
import net.greenjab.fixedminecraft.hud.HotbarCycler;
import net.greenjab.fixedminecraft.map_book.MapBookFilledProperty;
import net.greenjab.fixedminecraft.models.CustomModelLayers;
import net.greenjab.fixedminecraft.registries.CustomEntityModelLayerRegistry;
import net.greenjab.fixedminecraft.registries.EntityRendererRegistry;
import net.greenjab.fixedminecraft.screens.FletchingScreen;
import net.greenjab.fixedminecraft.screens.NewAnvilScreen;
import net.greenjab.fixedminecraft.screens.NewEnchantmentScreen;
import net.greenjab.fixedminecraft.registry.registries.MenuRegistry;
import net.greenjab.fixedminecraft.render.ChiseledBookRenderer;
import net.minecraft.client.Minecraft;
import net.minecraft.client.OptionInstance;
import net.minecraft.client.gui.screens.MenuScreens;
import net.minecraft.client.renderer.item.properties.conditional.ConditionalItemModelProperties;
import net.minecraft.network.chat.Component;

public class FixedMinecraftClient implements ClientModInitializer {
    public static float paleGardenFog = 0f;
    public static float voidFog = 0f;
    public static boolean fontLegend = false;

    public static OptionInstance<Boolean> itemArmorHud = OptionInstance.createBoolean("options.itemArmorHud", true);
    public static OptionInstance<Boolean> jabsFixedFog = OptionInstance.createBoolean("options.jabsFixedFog", true);
    public static OptionInstance<Boolean> villagersSpeak = OptionInstance.createBoolean("options.chat.villagersSpeak", true);

    @Override
    public void onInitializeClient() {

        ClientSyncHandler.init();

        MenuScreens.register(MenuRegistry.FLETCHING_SCREEN_HANDLER, FletchingScreen::new);
        MenuScreens.register(MenuRegistry.NEW_ENCHANTMENT_SCREEN_HANDLER, NewEnchantmentScreen::new);
        MenuScreens.register(MenuRegistry.NEW_ANVIL_SCREEN_HANDLER, NewAnvilScreen::new);

        HotbarCycler.register();

        ChiseledBookRenderer CBR = new ChiseledBookRenderer();
        HudElementRegistry.addLast(FixedMinecraft.id("chiseled_book"), CBR);

        ConditionalItemModelProperties.ID_MAPPER.put(FixedMinecraft.id("map_book/filled"), MapBookFilledProperty.CODEC);
        CustomModelLayers.onRegisterLayers();
        EntityRendererRegistry.registerEntityRenderer();
        CustomEntityModelLayerRegistry.registerEntityModelLayer();

        FabricLoader.getInstance().getModContainer("fixedminecraft").ifPresent(modContainer -> {
            ResourceManagerHelper.registerBuiltinResourcePack(
                    FixedMinecraft.id( "green_tweaks"),
                    modContainer,
                    Component.translatable("fixedminecraft.green_tweaks"),
                    ResourcePackActivationType.NORMAL
            );
            ResourceManagerHelper.registerBuiltinResourcePack(
                    FixedMinecraft.id( "recolourful_containers"),
                    modContainer,
                    Component.translatable("fixedminecraft.recolourful_containers"),
                    ResourcePackActivationType.NORMAL
            );
            ResourceManagerHelper.registerBuiltinResourcePack(
                    FixedMinecraft.id( "re_covered"),
                    modContainer,
                    Component.translatable("fixedminecraft.re_covered"),
                    ResourcePackActivationType.NORMAL
            );
            ResourceManagerHelper.registerBuiltinResourcePack(
                    FixedMinecraft.id( "almost_vanilla_potions"),
                    modContainer,
                    Component.translatable("fixedminecraft.almost_vanilla_potions"),
                    ResourcePackActivationType.NORMAL
            );
            ResourceManagerHelper.registerBuiltinResourcePack(
                    FixedMinecraft.id( "fixed_pacp"),
                    modContainer,
                    Component.translatable("fixedminecraft.fixed_pacp"),
                    ResourcePackActivationType.NORMAL
            );
        });
    }

    public static boolean usingCustomContainers() {
        return (Minecraft.getInstance().getResourcePackRepository().getSelectedPacks().stream().anyMatch(pack -> pack.location().id().toLowerCase().contains("recolourful_containers")));
    }
}

File: src/client/kotlin/net/greenjab/fixedminecraft/FixedMinecraftClient.kt
package net.greenjab.fixedminecraft

import net.fabricmc.api.ClientModInitializer
import net.fabricmc.fabric.api.blockrenderlayer.v1.BlockRenderLayerMap
import net.fabricmc.fabric.api.client.rendering.v1.HudRenderCallback
import net.fabricmc.fabric.api.resource.ResourceManagerHelper
import net.fabricmc.fabric.api.resource.ResourcePackActivationType
import net.fabricmc.loader.api.FabricLoader
import net.fabricmc.loader.api.ModContainer
import net.greenjab.fixedminecraft.map_book.MapBookFilledProperty
import net.greenjab.fixedminecraft.models.ModelLayers
import net.greenjab.fixedminecraft.network.ClientSyncHandler
import net.greenjab.fixedminecraft.registry.registries.BlockRegistry
import net.greenjab.fixedminecraft.render.PlayerLookHelper
import net.minecraft.client.MinecraftClient
import net.minecraft.client.gui.DrawContext
import net.minecraft.client.render.RenderLayer
import net.minecraft.client.render.RenderTickCounter
import net.minecraft.client.render.item.property.bool.BooleanProperties
import net.minecraft.item.ItemStack
import net.minecraft.item.Items
import net.minecraft.text.Text
import net.minecraft.util.Identifier

object FixedMinecraftClient : ClientModInitializer {
    var paleGardenFog = 0f
    override fun onInitializeClient() {

        ClientSyncHandler.init()

        BlockRenderLayerMap.INSTANCE.putBlocks(
            RenderLayer.getCutout(),
            BlockRegistry.COPPER_RAIL,
            BlockRegistry.EXPOSED_COPPER_RAIL,
            BlockRegistry.WEATHERED_COPPER_RAIL,
            BlockRegistry.OXIDIZED_COPPER_RAIL,
            BlockRegistry.WAXED_COPPER_RAIL,
            BlockRegistry.WAXED_EXPOSED_COPPER_RAIL,
            BlockRegistry.WAXED_WEATHERED_COPPER_RAIL,
            BlockRegistry.WAXED_OXIDIZED_COPPER_RAIL,
        )


        HudRenderCallback.EVENT.register(HudRenderCallback { drawContext: DrawContext, tickDeltaManager: RenderTickCounter ->
            val matrices = drawContext.matrices
            matrices.push()


            val client = MinecraftClient.getInstance()

            val book = PlayerLookHelper.getLookingAtBook(null)
            if (book != ItemStack.EMPTY) {

            val display = PlayerLookHelper.getBookText(book)
            for (i in display.indices) {
                val text = display[i]
                drawContext.drawText(
                    client.textRenderer,
                    text,
                    (client.window.scaledWidth / 2.0 - client.textRenderer.getWidth(text) / 2).toInt(),
                    (client.window.scaledHeight / 2.0 + 15 + (i * 10)).toInt(),
                    if ((book.item === Items.ENCHANTED_BOOK && i == 0)) 16777045 else 16777215,
                    true
                )
            }
            }

            matrices.pop()
        })

        BooleanProperties.ID_MAPPER.put(FixedMinecraft.id("map_book/filled"), MapBookFilledProperty.CODEC)
        ModelLayers.onRegisterLayers()

        FabricLoader.getInstance().getModContainer("fixedminecraft").ifPresent { modContainer: ModContainer? ->
            ResourceManagerHelper.registerBuiltinResourcePack(
                Identifier.of("fixedminecraft", "greentweaks"),
                modContainer,
                Text.of("Green Tweaks"),
                ResourcePackActivationType.DEFAULT_ENABLED
            )
        }
    }
}
